### PR TITLE
Nerf or HELL

### PR DESCRIPTION
**Pull request by taukausanake**
Ports /tg/ PR "Foam force fuck yeah! (#8451)". Includes most changes in there as some of them were already implemented with gun code I believe.

Problems:
* ~~Foam bullets shoot sideways... Need to figure out why or just turn the icon the proper direction.~~ Don't give projectiles direction icons. Now I think they just fire backwards.
* Foam darts do not have their icon showing up properly. Not sure why, everything is in one .dmi file and the gun icons work. When firing, dart shows but after hitting it disappears. Item is still there if you Alt-click the tile.
* ~~Miiiight~~ Do have two versions of the toy crossbow. "/obj/item/toy/crossbow" and "/obj/item/weapon/gun/projectile/shotgun/toy/crossbow". Both have items dependent on each, tommy gun and foam shotgun respectively. Someone tell me the best way to go about this as I might do something stupid otherwise.

It looks like nothing broke with the port. I think I might need a little help trouble'shooting' these ~~at NTSAM~~

### CODE REVIEW

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/obj/item/weapon/gun/projectile/automatic/toy
```
Comment (tigercat2000) [inline code/modules/projectiles/guns/projectile/toy.dm]: should really probably be it's own subtype of /obj/item/weapon/gun/projectile/automatic/toy

**Comment by FlattestGuitar** [comment]
I can sleep well now, knowing that we can pew pew foam darts properly now. :+1: x :100:

**Comment by FlattestGuitar** [comment]
Also, I think in general the sounds shouldn't be gunshots. There already are enough things to trigger sec PTSD and we don't need to add any more. Maybe just leave them on the nukie weapons for ultimate keks.

**Comment by Keilden** [comment]
Do they also cause the stun state when you fire without ammo?

**Comment by FlattestGuitar** [comment]
Nope, that would be silly.

**Comment by Fox-McCloud** [comment]
I will never understand why these can be purcahsed by nuke ops.

**Comment by FlattestGuitar** [comment]
It's a good alternative to the 380 soaps on war ops.

File: code/modules/projectiles/ammunition/ammo_casings.dm [hunk 2]
```diff
@@ -250,6 +250,43 @@
...
+/obj/item/ammo_casing/caseless/foam_dart/attackby(var/obj/item/A as obj, mob/user as mob, params)
+	..()
+	if (istype(A, /obj/item/weapon/screwdriver) && !modified)
+		modified = 1
+		BB.damage_type = BRUTE
+		icon_state = "foamdart_empty"
+		desc = "Its nerf or nothing! ...Although, this one doesn't look too safe."
+		user << "<span class='notice'>You pop the safety cap off of [src].</span>"
+	else if ((istype(A, /obj/item/weapon/pen)) && modified && !BB.contents.len)
+		user.drop_item()
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: if(!user.unEquip(A))
     return

File: code/modules/projectiles/ammunition/ammo_casings.dm [hunk 3]
```diff
@@ -250,6 +250,43 @@
 	caliber = "75"
 	projectile_type = "/obj/item/projectile/bullet/gyro"
 
+/obj/item/ammo_casing/caseless/foam_dart
+	name = "foam dart"
+	desc = "Its nerf or nothing! Ages 8 and up."
+	projectile_type = /obj/item/projectile/bullet/reusable/foam_dart
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: This needs to be a string path.

*sream

File: code/modules/projectiles/ammunition/ammo_casings.dm [hunk 4]
```diff
@@ -250,6 +250,43 @@
...
+		A.loc = BB
+		BB.damage = 5
+		BB.nodamage = 0
+		user << "<span class='notice'>You insert [A] into [src].</span>"
+	return
+
+/obj/item/ammo_casing/caseless/foam_dart/riot
+	name = "riot foam dart"
+	desc = "Who's smart idea was it to use toys as crowd control? Ages 18 and up."
+	projectile_type = /obj/item/projectile/bullet/reusable/foam_dart/riot
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: This needs to be a string path.

**Comment by NTSAM** [comment]
oh god you're a deity :+1: :heart:

File: code/modules/projectiles/ammunition/magazines.dm [hunk 5]
```diff
@@ -290,4 +298,50 @@
 /obj/item/ammo_box/magazine/m12g/dragon
 	name = "shotgun magazine (12g dragon's breath)"
 	icon_state = "m12gf"
-	ammo_type = "/obj/item/ammo_casing/shotgun/incendiary/dragonsbreath"
\ No newline at end of file
+	ammo_type = "/obj/item/ammo_casing/shotgun/incendiary/dragonsbreath"
+
+/obj/item/ammo_box/magazine/toy
+	name = "foam force META magazine"
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/magazines.dm]: String path.

File: code/modules/projectiles/ammunition/magazines.dm [hunk 6]
```diff
@@ -51,6 +51,14 @@
 	max_ammo = 4
 	multiload = 0
 
+/obj/item/ammo_box/magazine/internal/shot/toy
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/magazines.dm]: String path.

File: code/modules/projectiles/ammunition/magazines.dm [hunk 7]
```diff
@@ -290,4 +298,50 @@
...
+	max_ammo = 20
+
+/obj/item/ammo_box/magazine/toy/smgm45/update_icon()
+	..()
+	icon_state = "c20r45-[round(ammo_count(),2)]"
+
+/obj/item/ammo_box/magazine/toy/m762
+	name = "donksoft box magazine"
+	caliber = "foam_force"
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart/riot
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/magazines.dm]: string path

File: code/modules/projectiles/ammunition/magazines.dm [hunk 8]
```diff
@@ -290,4 +298,50 @@
...
+	max_ammo = 8
+	multiple_sprites = 2
+
+/obj/item/ammo_box/magazine/toy/pistol/riot
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart/riot
+
+/obj/item/ammo_box/magazine/toy/smgm45
+	name = "donksoft SMG magazine"
+	caliber = "foam_force"
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart/riot
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/magazines.dm]: string path

File: code/modules/projectiles/ammunition/magazines.dm [hunk 9]
```diff
@@ -290,4 +298,50 @@
...
+	name = "foam force SMG magazine"
+	icon_state = "smg9mm-20"
+	max_ammo = 20
+
+/obj/item/ammo_box/magazine/toy/smg/update_icon()
+	..()
+	icon_state = "smg9mm-[round(ammo_count(),5)]"
+
+/obj/item/ammo_box/magazine/toy/smg/riot
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart/riot
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/magazines.dm]: string path

File: code/modules/projectiles/ammunition/magazines.dm [hunk 10]
```diff
@@ -290,4 +298,50 @@
...
+
+/obj/item/ammo_box/magazine/toy/pistol
+	name = "foam force pistol magazine"
+	icon_state = "9x19p"
+	max_ammo = 8
+	multiple_sprites = 2
+
+/obj/item/ammo_box/magazine/toy/pistol/riot
+	ammo_type = /obj/item/ammo_casing/caseless/foam_dart/riot
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/magazines.dm]: string path

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+	item_state = "gun"
+	mag_type = /obj/item/ammo_box/magazine/toy/smg
+	fire_sound = 'sound/weapons/Gunshot_smg.ogg'
+	force = 0
+	throwforce = 0
+	burst_size = 3
+	can_suppress = 0
+	clumsy_check = 0
+	needs_permit = 0
+
```
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: Where's the handling so it doesn't eject casings?

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+	name = "foam force shotgun"
+	desc = "A toy shotgun with wood furniture and a four-shell capacity underneath. Ages 8 and up."
+	icon = 'icons/obj/guns/toy.dmi'
+	force = 0
+	throwforce = 0
+	origin_tech = null
+	mag_type = /obj/item/ammo_box/magazine/internal/shot/toy
+	clumsy_check = 0
+	needs_permit = 0
+
```
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: This doesn't have a process chamber...

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+	w_class = 2
+
+/obj/item/weapon/gun/projectile/automatic/c20r/toy
+	name = "donksoft SMG"
+	desc = "A bullpup two-round burst toy SMG, designated 'C-20r'. Ages 8 and up."
+	icon = 'icons/obj/guns/toy.dmi'
+	can_suppress = 0
+	needs_permit = 0
+	mag_type = /obj/item/ammo_box/magazine/toy/smgm45
+
```
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: or this one...

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+/obj/item/weapon/gun/projectile/automatic/l6_saw/toy
+	name = "donksoft LMG"
+	desc = "A heavily modified toy light machine gun, designated 'L6 SAW'. Ages 8 and up."
+	icon = 'icons/obj/guns/toy.dmi'
+	can_suppress = 0
+	needs_permit = 0
+	mag_type = /obj/item/ammo_box/magazine/toy/m762
```
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: or this one.

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+	fire_sound = 'sound/weapons/Gunshot.ogg'
+	can_suppress = 0
+	burst_size = 1
+	fire_delay = 0
+	action_button_name = null
+
+/obj/item/weapon/gun/projectile/automatic/toy/pistol/update_icon()
+	..()
+	icon_state = "[initial(icon_state)][chambered ? "" : "-e"]"
+
```
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: You didn't include the base type of `/obj/item/weapon/gun/projectile/automatic/toy/pistol/riot`

File: code/modules/projectiles/projectile/reusable.dm [hunk 16]
```diff
@@ -0,0 +1,42 @@
+/obj/item/projectile/bullet/reusable
+	name = "reusable bullet"
+	desc = "How do you even reuse a bullet?"
+	var/obj/item/ammo_casing/caseless/ammo_type = /obj/item/ammo_casing/caseless/
+
+/obj/item/projectile/bullet/reusable/on_hit(var/atom/target, var/blocked = 0)
+	if(!proj_hit)
```
Comment (Fox-McCloud) [inline code/modules/projectiles/projectile/reusable.dm]: this is handled differently on TG...any idea why or what makes it different here?

File: code/modules/research/designs/autolathe_designs.dm [hunk 17]
```diff
@@ -539,6 +539,14 @@
 	build_path = /obj/item/device/radio/electropack
 	category = list("hacked", "Tools")
 
+/datum/design/foam_dart
+	name = "Box of Foam Darts"
+	id = "foam_dart"
+	build_type = AUTOLATHE
+	materials = list("$metal" = 500)
+	build_path = /obj/item/ammo_box/foambox
+	category = list("initial", "Misc")
```
Comment (Fox-McCloud) [inline code/modules/research/designs/autolathe_designs.dm]: this won't show up, as we don't have a "Misc" category. You'll want full out `"Miscellaneous"`

**Comment by Fox-McCloud** [comment]
You messed up your .dme  Not idea what you did, there.

Also, this doesn't get rid of the old foam weaponss---those absolutely need to go.

File: code/modules/projectiles/ammunition/ammo_casings.dm [hunk 18]
```diff
@@ -250,6 +250,43 @@
 	caliber = "75"
 	projectile_type = "/obj/item/projectile/bullet/gyro"
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: For above, you need to implement the caseless icon update changes that were in the foam force PR.

**Comment by taukausanake** [comment]
Holy fuck never again will I expect a port to go smoothly enough. Good thing I'm only a single idiot. I'll get to work on everything in here as soon as I can

**Comment by NTSAM** [comment]
:gun: do iiitttttt >:(

**Comment by Fox-McCloud** [comment]
To other coders: 

Yes, our guncode is so awful as to require string paths to properly function.

**Comment by Spacemanspark** [comment]
I was going to port over foam force guns.
Then I started getting compile errors relating to the Die() proc.
So I said "Nope" and ran away screaming.

**Comment by TheDZD** [comment]
Fixing the sprite directional issues is easy, set the sprite to only have a single direction rather than 4.

**Comment by taukausanake** [comment]
@TheDZD Yeah I did that the moment I found it. Are projectiles supposed to be pointing downwards you think? It's hard to tell with lasers and bullets that are tiny little dots. Either way I'll find out soon. Back to work

**Comment by TheDZD** [comment]
Front of the projectile should be pointing up.

File: code/modules/projectiles/guns/projectile/toy.dm (hunk 11 above)
Comment (taukausanake) [inline code/modules/projectiles/guns/projectile/toy.dm]: process_chamber() wasn't found so I thought it was an unneeded proc. It looks like we use process_chambered() though?

File: code/modules/projectiles/ammunition/ammo_casings.dm (hunk 2 above)
Comment (taukausanake) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: I hope I added this in correctly

File: code/modules/projectiles/guns/projectile/toy.dm (hunk 14 above)
Comment (taukausanake) [inline code/modules/projectiles/guns/projectile/toy.dm]: Out of curiosity, when looking in code\modules\projectiles\ammunition\boxes.dm, ammo_type isn't a string path. Just wondering how they're different

File: code/modules/projectiles/guns/projectile/toy.dm (hunk 15 above)
Comment (taukausanake) [inline code/modules/projectiles/guns/projectile/toy.dm]: I don't even know if it was declared aside from that new() proc. Should I make one and edit the mag_type to riot?

File: code/modules/projectiles/ammunition/ammo_casings.dm (hunk 18 above)
Comment (taukausanake) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: No idea how I missed that

File: code/modules/projectiles/guns/projectile/toy.dm (hunk 14 above)
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: those are ammo boxes, used for reloading a gun---these are magazines, which are different (but a child of boxes for the sake of holding ammo).

**Comment by taukausanake** [comment]
Went through everything mentioned in the PR and replaced crossbow and tommy gun. Shotgun and crossbow do not shoot anymore. Need to find out why

**Comment by taukausanake** [comment]
Actually it seems none of them are working. What did I do?

File: code/modules/projectiles/ammunition/ammo_casings.dm [hunk 19]
```diff
@@ -245,11 +245,55 @@
...
+	if (istype(A, /obj/item/weapon/screwdriver) && !modified)
+		modified = 1
+		BB.damage_type = BRUTE
+		icon_state = "foamdart_empty"
+		desc = "Its nerf or nothing! ...Although, this one doesn't look too safe."
+		user << "<span class='notice'>You pop the safety cap off of [src].</span>"
+	else if ((istype(A, /obj/item/weapon/pen)) && modified && !BB.contents.len)
+		if (!user.unEquip(A))
+			return
+		user.drop_item()
```
Comment (Fox-McCloud) [inline code/modules/projectiles/ammunition/ammo_casings.dm]: you didn't get rid of drop_item.

**Comment by Fox-McCloud** [comment]
I really don't understand why you changed the base toy type--it messed everything up in terms of consistency. I know @tigercat2000 commented what they did, but with how those particular guns are handled, I don'tttt think it really matters.

File: code/datums/supplypacks.dm [hunk 20]
```diff
@@ -1407,6 +1407,25 @@ var/list/all_supply_groups = list(supply_emergency,supply_security,supply_engine
 	cost = 30 //Lots of very expensive items. You gotta pay up to look good!
 	containername = "formal-wear crate"
 
+/datum/supply_packs/misc/foamforce
+	name = "Foam Force Crate"
+	contains = list(/obj/item/weapon/gun/projectile/shotgun/toy,
+					/obj/item/weapon/gun/projectile/shotgun/toy,
+					/obj/item/weapon/gun/projectile/shotgun/toy,
+					/obj/item/weapon/gun/projectile/shotgun/toy,)
```
Comment (Fox-McCloud) [inline code/datums/supplypacks.dm]: why reduce this from 8 to 4?

Comment (taukausanake) [inline code/datums/supplypacks.dm]: I thought eight was too many. I'll change it back

**Comment by FalseIncarnate** [comment]
Merge Conflicts

**Comment by taukausanake** [comment]
I had to revert my commit and figure out what was going on. For some reason the process_chambered() proc makes the foam guns not fire. Commented them out and they work but they spew blank shells that really shouldn't exist. Need to find out why that's happening. Otherwise everything else should be fine

**Comment by taukausanake** [comment]
So I got all the dart guns to work at last. I don't like how I did it but I did somehow. I'm pretty sure there's supposed to be a better way but just declaring the proc with the changed var wasn't doing it. Couldn't find out why but maybe someone else can

File: code/modules/projectiles/guns/projectile/toy.dm [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+/obj/item/weapon/gun/projectile/automatic/l6_saw/toy
+	name = "donksoft LMG"
+	desc = "A heavily modified toy light machine gun, designated 'L6 SAW'. Ages 8 and up."
+	icon = 'icons/obj/guns/toy.dmi'
+	can_suppress = 0
+	needs_permit = 0
+	mag_type = "/obj/item/ammo_box/magazine/toy/m762"
+
+/obj/item/weapon/gun/projectile/automatic/l6_saw/toy/process_chambered(var/eject_casing = 0, var/empty_chamber = 1)
```
Comment (Fox-McCloud) [inline code/modules/projectiles/guns/projectile/toy.dm]: did...you really copy paste `process_chambered` for all these guns?

Comment (taukausanake) [inline code/modules/projectiles/guns/projectile/toy.dm]: IT'S THE ONLY THING THAT WORKED. :| If anything else did then I would have done that instead. Can someone help me find out why passing the proc with edited argument and ..() doesn't work?

**Comment by Fox-McCloud** [comment]
Are you going to continue to work on this or are ya going to leave it to someone else?

**Comment by taukausanake** [comment]
@Fox-McCloud I'm going to work on it more. I know I can get them to work better. I just don't understand why those procs weren't going through properly. But yeah, I'll get back on it now

**Comment by Fox-McCloud** [comment]
closing due to lack of updates.

**Comment by NTSAM** [comment]
B-but @taukausanake you were so close :(

**Comment by taukausanake** [comment]
Barely. I'll wait for updated gun code like others said

**Pull request closed without merging**